macros/generate_audit_report.sql

```diff
@@ -0,0 +1,95 @@
+        {{ exceptions.raise_compiler_error("Audit helper macro compare_all_columns requires primary_keys input.") }}
+    {% endif %}
+
+    {% if audit_helper_macro_name == 'compare_relation_columns' %}
+        {% if exclude_columns|length %}
+            {{ exceptions.raise_compiler_error("Audit helper macro compare_relation_columns does not accept exclude_columns input.") }}
+        {% elif primary_keys|length %}
+            {{ exceptions.raise_compiler_error("Audit helper macro compare_relation_columns does not accept primary_keys input.") }}
+        {% endif %}
+    {% endif %}
+
+    with
+
+    {% if execute %}
+
+    {% for m in model_list %}
+
+        {% set m_id = 'model.' ~ project_name ~ '.' ~ m %}
+        {% set m_details = graph["nodes"][m_id] %}
+
+    {{ m }}_results as (
+        {#- *---------------------------------------------
+            * This is where the audit_helper macro is used!
+        --------------------------------------------------* -#}
+
+        {% if identifiers[m] %}
+            {% set legacy_model_name = identifiers[m] %}
+        {% else %}
+            {% set legacy_model_name = m %}
+        {% endif %}
+
+        {% set compare_relation = adapter.get_relation(
+            database = compare_database,
+            schema = compare_schema,
+            identifier = legacy_model_name
+        ) -%}
+
+        {% set dbt_relation = adapter.get_relation(
+            database = m_details.database,
+            schema = m_details.schema,
+            identifier = m
+        ) -%}
+
+        {% set audit_helper_macro = audit_helper.get(audit_helper_macro_name) %}
+
+        {% set model_primary_key = primary_keys[m] %}
+
+        {% set model_exclude_columns = exclude_columns[m] %}
+
+        {% if not compare_relation %}
+
+            {{ exceptions.raise_compiler_error("Compare relation does not exist - confirm you have supplied the correct compare_database, compare_schema, and identifiers.") }} 
+
+        {% elif dbt_relation is none %}
+
+            {{ exceptions.raise_compiler_error("dbt model has not been materialized in the warehouse - try running a dbt build for model " ~ m) }} 
+        
+        {% else %}
+
+            {% do log("Compare relation (a) " ~ compare_relation ~ " to dbt model (b) " ~ dbt_relation, True) %}
+    
+            {% if audit_helper_macro_name != 'compare_relation_columns' %}
+                {{ audit_helper_macro(a_relation = compare_relation, b_relation = dbt_relation, primary_key = model_primary_key, exclude_columns = model_exclude_columns) }}
+            {% else %}
+                {{ audit_helper_macro(a_relation = compare_relation, b_relation = dbt_relation) }}
+            {% endif %}
+
+        {% endif %}
+
+    ),
+    {% endfor %}
+
+    final as (
+        {% for m in model_list %}
+        select 
+            '{{ m }}' as audit_model, 
+            * from {{ m }}_results
+        {% if not loop.last %}union all{% endif %}
```

Comment:
I feel like there are 2 ways to tackle this:

1. using `union all`
2. using `run_query()` for each model, storing the results and unioning them in Jinja/Agate ([see here](https://github.com/wireservice/agate/issues/759)). The results of the queries are only a  few lines so it should not create a memory problem when running dbt

I don't know what is the most performant and it will most likely depend on the Warehouse used and maybe the number of tables to compare as well.

e.g. BQ might run better with option 2. knowing that it doesn't like queries with many ctes/unions but for other warehouses, option 1 might just be much faster

Have you done some tests today on different warehouses and with different orders of tables to compare (e.g. 5, 20, 100)? And if not, would it make sense to spend the effort doing it?

Reply:
I have not done that sort of test, but agree that option 2 would likely be better for BigQuery. What about an option 3 where we materialize temporary tables, union, then drop the temp tables? Which are the warehouses that the audit_helper package is compatible with? Is it snowflake, redshift, bigquery, and postgres?

Reply:
The issue with option 3 would be that if one of the temp tables fails, then the macro will stop and we will never drop the temp tables that were created before the failure.

When retrieving results with `run_query()`, we never materialize anything in the warehouse, so it doesn't matter if something fails in the middle of the process. This `run_query()` approach works though only because the result sets are extremely small (just a few rows). If we had to return a bigger set, the temp tables would prevent loading the results in the memory of the machine running dbt.